cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/lineage/LineageServiceTest.java

```diff
@@ -38,73 +39,358 @@
+    lineageStore.addAccess(run1, dataset2, AccessType.WRITE, "metadata12", flowlet1);
 
-    // Lineage for D2 should be D3 -> P2 -> D2
-    Lineage lineage2 = lineageService.computeLineage(Id.DatasetInstance.from("default", "dataset2"), 500, 20000, 5);
-    Assert.assertNotNull(lineage2);
+    lineageStore.addAccess(run2, dataset2, AccessType.READ, "metadata22", flowlet2);
+    lineageStore.addAccess(run2, dataset1, AccessType.WRITE, "metadata21", flowlet2);
+    lineageStore.addAccess(run2, dataset3, AccessType.WRITE, "metadata23", flowlet2);
 
-    Assert.assertEquals(ImmutableSet.of(program2), lineage2.getPrograms());
-    Assert.assertEquals(ImmutableSet.of(datasetInstance2, datasetInstance3), lineage2.getData());
+    lineageStore.addAccess(run3, dataset3, AccessType.READ, "metadata33", flowlet3);
+    lineageStore.addAccess(run3, dataset4, AccessType.WRITE, "metadata34", flowlet3);
+
+    Lineage expectedLineage = new Lineage(
+      ImmutableSet.of(
+        new Relation(dataset2, program1, AccessType.WRITE, toSet(twillRunId(run1)), toSet(flowlet1)),
+        new Relation(dataset1, program1, AccessType.READ, toSet(twillRunId(run1)), toSet(flowlet1)),
+        new Relation(dataset1, program2, AccessType.WRITE, toSet(twillRunId(run2)), toSet(flowlet2)),
+        new Relation(dataset2, program2, AccessType.READ, toSet(twillRunId(run2)), toSet(flowlet2)),
+        new Relation(dataset3, program2, AccessType.WRITE, toSet(twillRunId(run2)), toSet(flowlet2)),
+        new Relation(dataset4, program3, AccessType.WRITE, toSet(twillRunId(run3)), toSet(flowlet3)),
+        new Relation(dataset3, program3, AccessType.READ, toSet(twillRunId(run3)), toSet(flowlet3))
+      ),
+      ImmutableSet.of(program1, program2, program3),
+      ImmutableSet.of(dataset1, dataset2, dataset3, dataset4)
+      );
+
+    // Lineage for D1
+    Assert.assertEquals(expectedLineage, lineageService.computeLineage(dataset1, 500, 20000, 100));
+
+    // Lineage for D2
+    Assert.assertEquals(expectedLineage, lineageService.computeLineage(dataset2, 500, 20000, 100));
+
+    // Lineage for D1 for one level D1 -> P1 -> D2 -> P2 -> D3
+    //                              |                 |
+    //                              |                 V
+    //                              |<-----------------
+    //
+    Lineage oneLevelLineage = lineageService.computeLineage(dataset1, 500, 20000, 1);
+
+    Assert.assertEquals(ImmutableSet.of(program1, program2), oneLevelLineage.getPrograms());
+    Assert.assertEquals(ImmutableSet.of(dataset1, dataset2, dataset3), oneLevelLineage.getData());
 
     Assert.assertEquals(
       ImmutableSet.of(
-        new Relation(datasetInstance2, program2, AccessType.WRITE, toSet(twillRunId(run2)), toSet(flowlet2)),
-        new Relation(datasetInstance3, program2, AccessType.READ, toSet(twillRunId(run2)), toSet(flowlet2))
+        new Relation(dataset2, program1, AccessType.WRITE, toSet(twillRunId(run1)), toSet(flowlet1)),
```

Comment:
why do we use twill run id here? or is it just the method name? In production it would be the cdap run id right?

Also, should we add a detailed javadoc to the ``Relation`` class? I think it would be helpful to explain what some of the members in that class stand for.